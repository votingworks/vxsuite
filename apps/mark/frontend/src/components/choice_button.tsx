import React from 'react';
import styled from 'styled-components';

import { Button, ButtonProps } from '@votingworks/ui';

import * as GLOBALS from '../config/globals';

interface Props
  extends ButtonProps,
    React.PropsWithoutRef<JSX.IntrinsicElements['button']> {
  choice: string;
  isSelected: boolean;
}

const StyledChoiceButton = styled('button').attrs(({ type = 'button' }) => ({
  role: 'option',
  type,
}))<Props>`
  position: relative;
  border-radius: 0.125rem;
  box-shadow: 0 0.125rem 0.125rem 0 rgba(0, 0, 0, 0.14),
    0 0.1875rem 0.0625rem -0.125rem rgba(0, 0, 0, 0.12),
    0 0.0625rem 0.3125rem 0 rgba(0, 0, 0, 0.2);
  background: ${({ isSelected }) => (isSelected ? '#028099' : '#FFFFFF')};
  cursor: pointer;
  padding: 0.5rem 0.5rem 0.5rem 4rem;
  @media (min-width: 480px) {
    padding: 1rem 1rem 1rem 4rem;
  }
  text-align: left;
  color: ${({ isSelected }) => (isSelected ? '#FFFFFF' : undefined)};
  transition: background 0.25s, color 0.25s;
  ::before {
    display: flex;
    align-items: center;
    justify-content: center;
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    border-right: 1px solid;
    border-color: ${({ isSelected }) =>
      isSelected ? '#028099' : 'rgb(211, 211, 211)'};
    border-radius: 0.125rem 0 0 0.125rem;
    background: #ffffff;
    width: 3rem;
    text-align: center;
    color: #028099;
    font-size: 2rem;
    font-weight: 700;
    content: '${({ isSelected }) => (isSelected ? GLOBALS.CHECK_ICON : '')}';
  }
`;

export function ChoiceButton({ choice, ...rest }: Props): JSX.Element {
  return (
    <Button
      {...rest}
      component={StyledChoiceButton}
      data-choice={choice}
      data-selected={rest.isSelected}
    />
  );
}
