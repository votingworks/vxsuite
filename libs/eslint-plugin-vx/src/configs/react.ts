import recommended = require('./recommended');

const jsExtensions = ['.js', '.jsx'];
const tsExtensions = ['.ts', '.tsx'];
const allExtensions = jsExtensions.concat(tsExtensions);

export = {
  ...recommended,
  extends: [
    ...recommended.extends
      .map((name) => (name === 'airbnb-base' ? 'airbnb' : name))
      .filter((name) => name !== 'plugin:prettier/recommended'),
    'airbnb/hooks',
    'plugin:react/recommended',
    'plugin:jsx-a11y/recommended',
    'plugin:prettier/recommended', // Enables eslint-plugin-prettier and displays prettier errors as ESLint errors. Make sure this is always the last configuration in the extends array.
  ],
  settings: {
    ...recommended.settings,
    react: {
      version: 'detect',
    },
  },
  rules: {
    ...recommended.rules,
    'jsx-a11y/label-has-associated-control': [
      'error',
      {
        controlComponents: ['Select'],
      },
    ],
    'react/jsx-filename-extension': ['error', { extensions: allExtensions }],
    'react/jsx-fragments': ['error', 'element'],
    'react/jsx-no-bind': 'off',
    'react/jsx-no-constructed-context-values': 'off',
    'react/jsx-no-useless-fragment': ['error', { allowExpressions: true }],
    'react/jsx-one-expression-per-line': 'off',
    'react/jsx-props-no-spreading': 'off',
    'react/jsx-wrap-multilines': 'off',
    'react/prop-types': 'off',
    'react/require-default-props': 'off',
  },
};
