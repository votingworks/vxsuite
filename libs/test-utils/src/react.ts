/* eslint-disable no-console -- this file monkey-patches `console` */

/**
 * Suppresses the "Can't perform a React state update on an unmounted component" warning by
 * modifying the `console.error` method to ignore the warning.
 *
 * This is a temporary workaround until we upgrade to React 18.
 */
export function suppressReact17UnmountedWarning(): void {
  const originalConsoleError = console.error;
  console.error = (...args: unknown[]) => {
    if (
      typeof args[0] === 'string' &&
      args[0].startsWith(
        `Warning: Can't perform a React state update on an unmounted component`
      )
    ) {
      return;
    }

    originalConsoleError(...args);
  };
}
