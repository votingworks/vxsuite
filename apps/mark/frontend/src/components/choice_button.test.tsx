import React from 'react';
import { render } from '@testing-library/react';

import { ChoiceButton } from './choice_button';

const onPress = jest.fn();

it('renders ChoiceButton', () => {
  const { container } = render(
    <ChoiceButton choice="foo" isSelected={false} onPress={onPress}>
      foo
    </ChoiceButton>
  );
  expect(container.firstChild).toMatchSnapshot();
});
