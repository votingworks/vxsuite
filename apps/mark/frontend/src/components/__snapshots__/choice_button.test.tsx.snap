// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders ChoiceButton 1`] = `
.c0 {
  position: relative;
  border-radius: 0.125rem;
  box-shadow: 0 0.125rem 0.125rem 0 rgba(0,0,0,0.14), 0 0.1875rem 0.0625rem -0.125rem rgba(0,0,0,0.12), 0 0.0625rem 0.3125rem 0 rgba(0,0,0,0.2);
  background: #FFFFFF;
  cursor: pointer;
  padding: 0.5rem 0.5rem 0.5rem 4rem;
  text-align: left;
  -webkit-transition: background 0.25s,color 0.25s;
  transition: background 0.25s,color 0.25s;
}

.c0::before {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  border-right: 1px solid;
  border-color: rgb(211,211,211);
  border-radius: 0.125rem 0 0 0.125rem;
  background: #ffffff;
  width: 3rem;
  text-align: center;
  color: #028099;
  font-size: 2rem;
  font-weight: 700;
  content: '';
}

@media (min-width:480px) {
  .c0 {
    padding: 1rem 1rem 1rem 4rem;
  }
}

<button
  class="c0"
  data-choice="foo"
  data-selected="false"
  role="option"
  type="button"
>
  foo
</button>
`;
